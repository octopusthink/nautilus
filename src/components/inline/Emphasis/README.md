An `<Emphasis>` is used to mark a short span of text within body copy as having a light stress.

```jsx
import { Paragraph } from '@octopusthink/nautilus';

<Paragraph>
  He wanted to eat a peanut butter cup, but he was <Emphasis>super</Emphasis> allergic to peanuts and may have perished in the attempt.
</Paragraph>
```

## Usage

Use an `<Emphasis>` component where you would usually use a `<em>` tag in HTML.

### Use this component for...

- Indicating stress or a shift in intonation in a passage of text.
- A piece of text requiring "local" emphasis (emphasis when reading the passage).
- Distinguishing titles of books, articles, and other works. 

### Don't use this component for...

- A heading or subheading. Use a `<Heading>` or `<Subtitle>` instead.
- Highlighting a navigational link. Use a `<Link>` instead.
- Helping users who are skimming to pick out key phrases and concepts. Use a `<Strong>` instead.
- A piece of text requiring "global" emphasis (emphasis when scanning the page as a whole). Use a `<Strong>` instead.

## Appearance

`<Emphasis>` spans are styled consistently throughout and have less visual weight than `<Strong>` spans. Colours are adjustable in your theme.

```jsx
import { Paragraph, Strong } from '@octopusthink/nautilus';

<Paragraph>
  The puppies <Strong>chased their tails</Strong> <Emphasis>almost</Emphasis> every single afternoon.
</Paragraph>
```

## Accessibility

Use care when emphasising text. Only emphasise a few words at a time. Some users may struggle to read italicised text, so only use this when necessary to convey additional information.
