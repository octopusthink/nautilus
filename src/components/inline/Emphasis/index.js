import { css } from '@emotion/core';
import styled from '@emotion/styled';
import PropTypes from 'prop-types';
import React, { forwardRef } from 'react';

export const Emphasis = forwardRef((props, ref) => {
  const { children, ...otherProps } = props;

  return (
    <em ref={ref} {...otherProps}>
      {children}
    </em>
  );
});

export const styles = (props) => {
  const { theme } = props;

  return css`
    background: ${theme.colors.text.emphasisBackground};
    color: ${theme.colors.text.emphasis};
    font-style: italic;
  `;
};

Emphasis.defaultProps = {
  children: undefined,
};

Emphasis.propTypes = {
  /** @ignore */
  children: PropTypes.node,
};

export const { defaultProps, propTypes } = Emphasis;

Emphasis.displayName = 'Emphasis';

export default styled(Emphasis)(styles);
