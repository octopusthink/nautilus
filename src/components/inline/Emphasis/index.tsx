import { css } from '@emotion/core';
import React, { FunctionComponent } from 'react';

import { NautilusComponentProps } from '../../types';
import { useTheme } from '../../../themes';

interface Props
  extends React.PropsWithoutRef<JSX.IntrinsicElements['em']>,
    NautilusComponentProps {}

const Emphasis: FunctionComponent<Props> = ({ children, unstyled, ...otherProps }: Props) => {
  const theme = useTheme();

  return (
    <em
      css={
        unstyled
          ? undefined
          : css`
              background: ${theme.colors.text.emphasisBackground};
              color: ${theme.colors.text.emphasis};
              font-style: italic;
            `
      }
      {...otherProps}
    >
      {children}
    </em>
  );
};

export default Emphasis;
