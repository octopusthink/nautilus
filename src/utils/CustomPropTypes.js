import { Children } from 'react';

const allowedChildren = (...types) => {
  return (propValue, key, componentName) => {
    let error = null;

    Children.forEach(propValue[key], (child) => {
      if (!types.includes(child.type)) {
        error = new Error(
          `${componentName} component only accepts the following components as children: ${types
            .map((element) => element.displayName || element)
            .join(', ')}.`,
        );
      }
    });

    return error;
  };
};

export default {
  allowedChildren,
};
