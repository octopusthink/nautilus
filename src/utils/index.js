export { default as CustomPropTypes } from './CustomPropTypes';
