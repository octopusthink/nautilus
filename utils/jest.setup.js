// import axios from 'axios';
// import AxiosMockAdapter from 'axios-mock-adapter';
// import { JSDOM } from 'jsdom';
// eslint-disable-next-line import/no-extraneous-dependencies
import 'jest-axe/extend-expect';
// eslint-disable-next-line import/no-extraneous-dependencies
import '@testing-library/jest-dom/extend-expect';

global.window.matchMedia = jest.fn().mockImplementation((query) => ({
  matches: false,
  media: query,
  onchange: null,
  addListener: jest.fn(),
  removeListener: jest.fn(),
}));

jest.spyOn(global.console, 'error');
jest.spyOn(global.console, 'warn');
jest.spyOn(global.console, 'log');
jest.spyOn(global.console, 'info');
jest.spyOn(global.console, 'debug');

beforeEach(() => {
  global.console.error.mockClear();
  global.console.warn.mockClear();
  global.console.log.mockClear();
  global.console.info.mockClear();
  global.console.debug.mockClear();
});
