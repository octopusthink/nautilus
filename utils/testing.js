// We only use react-testing-library in development, so ignore this linting
// error.
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createMemoryHistory } from 'history';
import PropTypes from 'prop-types';
import React from 'react';
import { Router } from 'react-router-dom';

import { Nautilus } from 'components';
import { nautilus as nautilusDefaultTheme, themePropTypes } from 'themes';

// Wrap `react-testing-library`'s `render` function with our providers.
const NautilusProviders = ({ children, theme }) => {
  return <Nautilus theme={theme}>{children}</Nautilus>;
};

NautilusProviders.defaultProps = {
  children: undefined,
  theme: nautilusDefaultTheme,
};

NautilusProviders.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** @ignore */
  theme: themePropTypes,
};

const customRender = (ui, options) => {
  return render(ui, { wrapper: NautilusProviders, ...options });
};

// this is a handy function that I would utilize for any component
// that relies on the router being in context
export const renderWithRouter = (
  ui,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  } = {},
) => {
  return {
    ...customRender(<Router history={history}>{ui}</Router>),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
  };
};

export const muteConsole = ({ times, type } = {}) => {
  Array.from({ length: times }).forEach(() => {
    global.console[type].mockImplementationOnce(() => jest.fn());
  });
};

// Export jest-axe helpers.
export { axe } from 'jest-axe';

// Re-export everything else as normal.
export * from '@testing-library/react';

// override render method
export { customRender as render };
