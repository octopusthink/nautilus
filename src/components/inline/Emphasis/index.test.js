import React from 'react';

import { axe, render } from 'utils/testing';

import Emphasis from '.';

describe('Emphasis', () => {
  it('should render an <em> tag', () => {
    const { container } = render(<Emphasis>Hello</Emphasis>);

    expect(container.firstChild.tagName).toEqual('EM');
  });

  it('should output its children', () => {
    const { getByTestId } = render(
      <Emphasis>
        <span data-testid="child" />
      </Emphasis>,
    );

    expect(getByTestId('child')).toBeDefined();
  });

  it('should accept and pass through other props', () => {
    const { getByTestId } = render(
      <Emphasis className="custom-class" data-testid="myText">
        hello
      </Emphasis>,
    );

    expect(getByTestId('myText').classList).toContain('custom-class');
  });

  it('should match styles', () => {
    const { container } = render(<Emphasis>hello</Emphasis>);

    expect(container).toMatchSnapshot();
  });

  describe('accessibility', () => {
    it('should pass aXe tests', async () => {
      const { container } = render(<Emphasis>hello</Emphasis>);

      expect(await axe(container.innerHTML)).toHaveNoViolations();
    });
  });
});
