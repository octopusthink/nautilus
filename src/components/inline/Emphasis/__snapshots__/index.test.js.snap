// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emphasis should match styles 1`] = `
.emotion-0 {
  background: transparent;
  color: #666c76;
  font-style: italic;
}

<div>
  <em
    class="emotion-0"
  >
    hello
  </em>
</div>
`;
